keps/sig-node/4112-passdown-resources-to-cri/README.md

```diff
@@ -0,0 +1,1054 @@
+container resource requests to be able to make optimal, platform specific,
+resource allocations. Some of the resources may be handled outside cgroups and
+be container runtime specific details.
+
+#### Story 3
+
+As a cluster administrator, I want to install an NRI plugin that does
+customized resource handling. I run kubelet with CPU manager and memory manager
+disabled. Instead I use my NRI plugin to do customized resource allocation
+(e.g. cpu and memory pinning). To do that properly I need the actual resource
+requests and limits requested by the user.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+The proposal only adds new informantional data to the CRI API between kubelet
+and the container runtime with no user-visible changes which mitigates possible
+risks considerably.
+
+Data duplication/inconsistency with native resources could be considered a risk
+as those are passed down to CRI both as "raw" requests and limits and as
+"translated" resource control parameters (like cpu shares oom scoring etc). But
+this should be largely mitigated by code reviews and unit tests.
+
+## Design Details
+
+The proposal is that kubelet discloses full resources information from the
+PodSpec to the container runtime. This is accomplished by extending the
+ContainerConfig, UpdateContainerResourcesRequest and PodSandboxConfig messages
+of the CRI API.
+
+With this information, the runtime can for example do detailed resource
+allocation so that CPU, memory and other resources for each container are
+optimally aligned.
+
+The resource information is included in PodSandboxConfig so that the runtime
+can see the full picture of Pod's resource usage at Pod creation time, for
+example enabling more holistic resource allocation and thus better
+interoperability between containers inside the Pod.
+
+### CRI API
+
+#### ContainerConfig
+
+The ContainerConfig message (used in CreateContainer request) is extended to
+contain unmodified resource requests from the PodSpec.
+
+```diff
++import "k8s.io/apimachinery/pkg/api/resource/generated.proto";
+
+ message ContainerConfig {
+ 
+ ...
+ 
+     // Configuration specific to Windows containers.
+     WindowsContainerConfig windows = 16;
+ 
+     // CDI devices for the container.
+     repeated CDIDevice CDI_devices = 17;
++
++    // Kubernetes resource spec of the container
```

Comment:
Hmm..but isn't the main use case to inform the runtime at podsandbox creation time what/how much resources to allocate? Having it in the podsandbox meets the purpose already. What'd the runtime do when getting this additional info?

Reply:
@yujuhong as said, it's for consistency/symmetry. Somewhat similar to PodSandboxConfig being passed on each CreateContainerRequest "just for easy reference". The may use that information for resource optimization when the container is created. But now, writing this I realize that the container resources are already in PodSandboxConfig so it could take it from there, if needed.

Reply:
everything we add is a contract we need to maintain, making it for consistency/symmetry reasons seems very expensive

Reply:
>  But now, writing this I realize that the container resources are already in PodSandboxConfig so it could take it from there, if needed.

this comment is still not resolved, do we need only `PodSandboxConfig` or do we need `ContainerConfig` too ... the symmetry argument does not seem to hold much weight though

Reply:
@kad to clarify, I'm not evaluating if technically is needed or not, or pushing back, is just that the way it is phrased it is not clear, doing something because of API for symmetry is a bit vague, don't you think?

Having at least some sentences explaining what you say, to understand the motivation why we are doing it, I think is important, at the end of the day, KEPs serve as the documentation of features and if we omit these details after a few years none of us will remember why we did this

Reply:
@aojea make sense.
@marquiz  can you please move bits of the comments to KEP text to clarify usage and reasoning for those fields?

Reply:
, appreciate it, +1 on my side, specially with the angle of the new work on pod mutability .
thanks